crates/proc_macro_api/src/process.rs

```diff
@@ -5,50 +5,37 @@ use std::{
     ffi::{OsStr, OsString},
     io::{self, BufRead, BufReader, Write},
     path::{Path, PathBuf},
-    process::{Child, Command, Stdio},
-    sync::{Arc, Weak},
+    process::{Child, ChildStdin, ChildStdout, Command, Stdio},
+    sync::Mutex,
 };
 
-use crossbeam_channel::{bounded, Receiver, Sender};
 use stdx::JodChild;
 
 use crate::{
     msg::{ErrorCode, Message, Request, Response, ResponseError},
     rpc::{ListMacrosResult, ListMacrosTask, ProcMacroKind},
 };
 
-#[derive(Debug, Default)]
-pub(crate) struct ProcMacroProcessSrv {
-    inner: Weak<Sender<Task>>,
-}
-
 #[derive(Debug)]
-pub(crate) struct ProcMacroProcessThread {
-    // XXX: drop order is significant
-    sender: Arc<Sender<Task>>,
-    handle: jod_thread::JoinHandle<()>,
+pub(crate) struct ProcMacroProcessSrv {
+    process: Mutex<Process>,
+    stdio: Mutex<(ChildStdin, BufReader<ChildStdout>)>,
```

Comment:
Repeated `Mutex`es are a smell -- it usually means that synchronization was added on the as-needed basis, as opposed to following some overarching concurrency principle. 

A specific problem here is that it's now possible to lock them in different orders, leading to the deadlock. 

I suggest pushing the `Mutex` upwards, to the caller, to the `ProcMacroProcessExpander` (I guess that's a second smell -- you have a mutex in one place, but actual resource sharing in another place). 

It also makes sense to document, on that field with Mutex, what we believe our overarching concurrency principle is :D 

I guess it's something like this:

All proc macros are expanded sequentially. That means that concurrent salsa requests may block each other when expanding proc macros, which is unfortunate, but simple and good enough for the time being. 



Having written this, I think I can see why we might want to have an extra thread here in the future -- we can make the process on the other side multithreaded, such that it can work on several requests simultaneously. That way, in rust-analyzer, concurrent queries won't be blocking each other. To make this work, we'd need a coordinating thread which would select between accepting a new request a reading a response to one of the several currently-outstanding requests.

Reply:
We did have _some_ support for concurrent requests when this was implemented :grimacing:.